### PR TITLE
[ZEPPELIN-1086] Auto Completion for Interpreter.

### PR DESCRIPTION
**Pull request by astroshim**
### What is this PR for?
This PR is to support auto-completion for interpreter.

### What type of PR is it?
Improvement 

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-1086

### How should this be tested?
Type ```% + 'Ctrl+.' ``` in your paragraph.
please refer to screen shot.

### Screenshots (if appropriate)
o

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by astroshim** [comment]
re-trigger CI

**Comment by AhyoungRyu** [comment]
@astroshim Awesome! :+1:  
I'm not sure where can be the proper place, it would be better users can know this feature's existence since it's really useful. How about adding this info in [here](https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/components/modal-shortcut/modal-shortcut.html#L211)? This is just my personal opinion :)

**Comment by astroshim** [comment]
@AhyoungRyu Thank you for your review!
I think it's better to make zeppelin manual on zeppelin homepage than adding shortcut info page. What do you think?

**Comment by AhyoungRyu** [comment]
@astroshim Oh sorry for my late response. Yeah that would be better 👍

**Comment by astroshim** [comment]
@bzz @Leemoonsoo please review.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java [hunk 1]
```diff
@@ -202,11 +201,26 @@ public Interpreter getCurrentRepl() {
...
+    for (InterpreterSetting intp: factory.getInterpreterSettings(note.getId())){
+      completion.add(new InterpreterCompletion(intp.getGroup(), intp.getGroup()));
+    }
+    return completion;
+  }
+
   public List<InterpreterCompletion> completion(String buffer, int cursor) {
+    String lines[] = buffer.split(System.getProperty("line.separator"));
+    if (lines.length > 0
+      && (lines[0].startsWith("%") && cursor <= lines[0].trim().length())) {
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: Code text can follow after interpreter selection without line.separator. For example

```sql
%sql select * from table
```
In this case, `getInterpreterCompletion()` will be invoked even if cursor is in somewhere `select * from table`. isn't it?

Comment (astroshim) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: You're right. I missed that point. Thank you pointing out.

Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: space or new line could be separator in this case?

Comment (astroshim) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Paragraph.java]: yes. It should be separated space or new line.

**Comment by astroshim** [comment]
@Leemoonsoo As you can see the following screenshot, I fixed bug and changed completion results.

o

**Comment by bzz** [comment]
@astroshim looks awesome, thank you for implementing!

How do you think, would it improve user experience with this feature if we could remove from suggestion list things, that are not interpreters? Like function names, etc. And how hard would that be to implement?

**Comment by astroshim** [comment]
@bzz Thank you for reviewing. 
You mean that blocking the keyword and local meta values?

**Comment by bzz** [comment]
What I mean is - as a user, when I auto-complete the interpreter name, I would expect a suggestion list that contains only interpreter names and nothing else.

Right now list of suggestion includes things like "NoSuchMethod" and `AbstractMethodError`.

If it is hard to remove such things from the suggestions list in this case - let's keep them. But if that is something we could implement in this PR - that would be nice.

How do you think?

**Comment by astroshim** [comment]
@bzz I totally agree with you so I did try to remove things you mentioned but it's not easy.
Let me try one more. Thanks!

**Comment by bzz** [comment]
Got it! Thanks a lot for looking into it.

Let me know when you think its ready and I will try again. You did great job already, making it auto-complete names so in case it's too hard - we could merge as it is now, its already very useful!

**Comment by astroshim** [comment]
@bzz Thank you kindly for telling like that. 
What do you think about tagging meta value than removing keywords like following image? 
image

I think it looks more nicer than removing keywords.
If it is okay, let me make it at the another PR.

**Comment by bzz** [comment]
Sounds as a great plan to me!

Merging this one then, if there is no further discussion

**Comment by astroshim** [comment]
Ping

**Comment by bzz** [comment]
Sorry, it took me a while to get back. Done!

**Comment by minahlee** [comment]
@astroshim this PR causes compile error in current master. Do you mind to create another PR to fix it?

**Comment by astroshim** [comment]
@minahlee Thank you for addressing that. I'll take a look.

**Comment by astroshim** [comment]
@minahlee Sorry for bothering you. I just made PR for fixing compile error.

**Pull request closed without merging**